### PR TITLE
Fix sequence numbers in profiler output

### PR DESCRIPTION
**Pull request by ilia-cher**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42565 Fix sequence numbers in profiler output**

Summary:
After recent changes to the record function we record more
ranges in profiler output and also keep emitting sequence numbers for
all ranges.

Sequence numbers are used by external tools to correlate forward
and autograd ranges and with many ranges having the same sequence number
it becomes impossible to do this.

This PR ensures that we set sequence numbers only for the top-level
ranges and only in case when autograd is enabled.

Test Plan:
nvprof -fo trace.nvvp --profile-from-start off python test_script.py
test_script
https://gist.github.com/ilia-cher/2baffdd98951ee2a5f2da56a04fe15d0
then examining ranges in nvvp

Differential Revision: [D22938828](https://our.internmc.facebook.com/intern/diff/D22938828)

### CODE REVIEW

**Comment by mcarilli** [comment]
Should close https://github.com/pytorch/pytorch/issues/42564.  By pure coincidence, the issue number is off by one from the PR number, which, considering the subject, proves the universe has a sense of humor.

**Comment by ilia-cher** [comment]
added test ( @ngimel )

**Review by dzhulgakov — COMMENTED** [review]
Looks good, will let other folks to take a look too.

Do we have first-class support for folding backwards into aggregation btw? I think that'd be useful as an option in `profile()`'s output

File: test/test_autograd.py [hunk 1]
```diff
@@ -2705,6 +2705,52 @@ def test_profiler(self):
...
+            s.backward()
+        # expecting aten::add, aten::sum to have the sequence numbers,
+        # expecting the corresponding backward nodes to have the same numbers
+        # as the forward ops
+        add_seq_nr = -1
+        sum_seq_nr = -1
+        found_add = found_sum = False
+        found_bwd_add = found_bwd_sum = False
+        found_empty = False
+        for e in p.function_events:
```
Comment (ngimel) [inline test/test_autograd.py:2723]: I assume this is after filtering, so you won't have multiple function events with the same name?

File: torch/csrc/autograd/function.h [hunk 2]
```diff
@@ -91,7 +91,7 @@ struct TORCH_API Node : std::enable_shared_from_this<Node> {
   /// a (currently THE) hint to prioritization in the backward() pass, with
   /// higher sequence numbers prioritized before lower sequence numbers.
   explicit Node(
-      uint64_t sequence_nr,
+      int64_t sequence_nr,
```
Comment (ngimel) [inline torch/csrc/autograd/function.h:94]: why did you need int64_t here? Normally autograd should not get negative seq_nrs?

**Comment by ngimel** [comment]
> Do we have first-class support for folding backwards into aggregation btw? I think that'd be useful as an option in `profile()`'s output

Not yet, but having sequence_nr as part of function_event is a first step towards that (before it was available only via nvtx labels, so only pyprof was using it)

File: test/test_autograd.py (hunk 1 above)
Comment (ilia-cher) [reply test/test_autograd.py:2723]: yes, this issue with double logging is not visible here, because its after filtering;
I also err'd on the side of failing a test, so added extra checks.

File: torch/csrc/autograd/function.h (hunk 2 above)
Comment (ilia-cher) [reply torch/csrc/autograd/function.h:94]: it's not necessary here, will upd

**Comment by ilia-cher** [comment]
comment and lint

**Review by ngimel — APPROVED** [review]
